### PR TITLE
[ITBL-4897] Improve FCM support, remove GCM

### PR DESCRIPTION
**Pull request by vbabenkoru**
Ready

### CODE REVIEW

File: iterableapi/src/main/java/com/iterable/iterableapi/IterablePushRegistration.java [hunk 1]
```diff
@@ -4,142 +4,117 @@
...
-            } else if (iterablePushRegistrationData.pushRegistrationAction == IterablePushRegistrationData.PushRegistrationAction.DISABLE) {
-                pushRegistrationObject = getDeviceToken(iterablePushRegistrationData.projectNumber, iterablePushRegistrationData.messagingPlatform, iterablePushRegistrationData.iterableAppId, false);
-                if (pushRegistrationObject != null) {
-                    IterableApi.sharedInstance.disablePush(pushRegistrationObject.token);
+    protected Void doInBackground(IterablePushRegistrationData... params) {
+        IterablePushRegistrationData iterablePushRegistrationData = params[0];
+        if (iterablePushRegistrationData.pushIntegrationName != null) {
+            PushRegistrationObject pushRegistrationObject = getDeviceToken();
+            if (pushRegistrationObject != null) {
+                if (iterablePushRegistrationData.pushRegistrationAction == IterablePushRegistrationData.PushRegistrationAction.ENABLE) {
```
Comment (davidtruong) [inline iterableapi/src/main/java/com/iterable/iterableapi/IterablePushRegistration.java]: It's possible that the user upgrades to FCM and then calls disable.
That means that we should also call disableOldDeviceIfNeeded from the disable case in order to disable the old GCM token. (move disableOldDeviceIfNeeded outside the if/if else block)

File: iterableapi/src/main/java/com/iterable/iterableapi/IterableFirebaseMessagingService.java [hunk 2]
```diff
@@ -0,0 +1,82 @@
...
+public class IterableFirebaseMessagingService extends FirebaseMessagingService {
+
+    static final String TAG = "itblFCMMessagingService";
+
+    @Override
+    public void onMessageReceived(RemoteMessage remoteMessage) {
+        if (remoteMessage.getData().size() > 0) {
+            Map<String,String> messageData = remoteMessage.getData();
+            handlePushReceived(messageData);
+            Log.d(TAG, "Message data payload: " + remoteMessage.getData());
```
Comment (davidtruong) [inline iterableapi/src/main/java/com/iterable/iterableapi/IterableFirebaseMessagingService.java]: Use the IterableLogger

File: iterableapi/src/main/java/com/iterable/iterableapi/IterableFirebaseMessagingService.java [hunk 3]
```diff
@@ -0,0 +1,82 @@
...
+        }
+
+        // Check if message contains a notification payload.
+        if (remoteMessage.getNotification() != null) {
+            Log.d(TAG, "Message Notification Body: " + remoteMessage.getNotification().getBody());
```
Comment (davidtruong) [inline iterableapi/src/main/java/com/iterable/iterableapi/IterableFirebaseMessagingService.java]: Use the IterableLogger

File: iterableapi/src/main/java/com/iterable/iterableapi/IterablePushRegistration.java [hunk 4]
```diff
@@ -4,142 +4,117 @@
 import android.content.SharedPreferences;
 import android.os.AsyncTask;
 
-import com.google.android.gms.gcm.GoogleCloudMessaging;
-import com.google.android.gms.iid.InstanceID;
+//import com.google.android.gms.gcm.GoogleCloudMessaging;
```
Comment (davidtruong) [inline iterableapi/src/main/java/com/iterable/iterableapi/IterablePushRegistration.java]: Remove commented imports

File: iterableapi/src/main/java/com/iterable/iterableapi/IterablePushRegistration.java [hunk 5]
```diff
@@ -4,142 +4,117 @@
...
-        }
-
-        String deviceToken = null;
-        if (pushRegistrationObject != null) {
-            deviceToken = pushRegistrationObject.token;
+            IterableLogger.e("IterablePush", "iterablePushRegistrationData has not been specified");
         }
-
-        return deviceToken;
+        return null;
```
[the remark this answers is not in the document]
Comment (vbabenkoru) [reply iterableapi/src/main/java/com/iterable/iterableapi/IterablePushRegistration.java]: The type is `Void`, so it must return an object (`null`).

File: iterableapi/src/main/java/com/iterable/iterableapi/IterablePushRegistration.java (hunk 1 above)
Comment (vbabenkoru) [reply iterableapi/src/main/java/com/iterable/iterableapi/IterablePushRegistration.java]: If they upgrade, they'll call `registerForPush` at some point to register the new token anyway.

File: iterableapi/src/main/java/com/iterable/iterableapi/IterablePushRegistration.java (hunk 5 above)
[the remark this answers is not in the document]
Comment (davidtruong) [reply iterableapi/src/main/java/com/iterable/iterableapi/IterablePushRegistration.java]: Got it it was a capital `Void`.

File: iterableapi/src/main/java/com/iterable/iterableapi/IterablePushRegistration.java (hunk 1 above)
Comment (davidtruong) [reply iterableapi/src/main/java/com/iterable/iterableapi/IterablePushRegistration.java]: My thought was about the old token. What happens if they upgrade to FCM and want to disable pushes on the device without ever attempting to register a new token?

Comment (virusman) [reply iterableapi/src/main/java/com/iterable/iterableapi/IterablePushRegistration.java]: Yeah, that makes sense. I’ll move this call.

Comment (vbabenkoru) [reply iterableapi/src/main/java/com/iterable/iterableapi/IterablePushRegistration.java]: Yeah, that makes sense. I’ll move this call.

**Pull request merged**